node/npm_dependencies.ts

```diff
@@ -0,0 +1,147 @@
+
+      if (isLocalImport) {
+        return
+      }
+
+      const isRemoteURLImport = args.path.startsWith('https://') || args.path.startsWith('http://')
+
+      if (!isRemoteURLImport) {
+        specifiers.add(args.path)
+      }
+
+      // At this point we know we're not dealing with a local import, so we're
+      // about to leave the boundaries of user code. We mark the specifier as
+      // external, because we're not interested in traversing the entire module
+      // tree — i.e. if user code imports module `foo` and that imports `bar`,
+      // we only want to add `foo` to the list of specifiers, since the whole
+      // module — including its dependencies like `bar` — will be bundled.
+      return { external: true }
+    })
+  },
+})
+
+export const vendorNPMSpecifiers = async (basePath: string, functions: string[], directory?: string) => {
+  const specifiers = new Set<string>()
+
+  // The directories that esbuild will use when resolving Node modules. We must
+  // set these manually because esbuild will be operating from a temporary
+  // directory that will not live inside the project root, so the normal
+  // resolution logic won't work.
+  const nodePaths = [path.join(basePath, 'node_modules')]
+
+  // Do a first pass at bundling to gather a list of specifiers that should be
+  // loaded as npm dependencies, because they either use the `npm:` prefix or
+  // they are bare specifiers. We'll collect them in `specifiers`.
+  await build({
+    bundle: true,
+    entryPoints: functions,
+    nodePaths,
+    platform: 'node',
+    plugins: [getDependencyTrackerPlugin(specifiers)],
+    write: false,
+  })
+
+  // If we found no specifiers, there's nothing left to do here.
+  if (specifiers.size === 0) {
+    return
+  }
+
+  // We need to create some files on disk, which we don't want to write to the
+  // project directory. If a custom directory has been specified, which happens
+  // only in tests, we use it. Otherwise, create a random temporary directory.
+  const temporaryDirectory = directory ? { path: directory } : await tmp.dir()
+
+  // To bundle an entire module and all its dependencies, we create a stub file
+  // where we re-export everything from that specifier. We do this for every
+  // specifier, and each of these files will be the entry points to esbuild.
+  const ops = await Promise.all(
+    [...specifiers].map(async (specifier, index) => {
+      const code = `export { default } from "${specifier}"; export * from "${specifier}";`
+      const filePath = path.join(temporaryDirectory.path, `stub-${index}.js`)
+
+      await fs.writeFile(filePath, code)
+
+      return { filePath, specifier }
+    }),
+  )
+  const entryPoints = ops.map(({ filePath }) => filePath)
+
+  // Bundle each of the stub files we've created. We'll end up with a compiled
+  // version of each of the stub files, plus any chunks of shared code between
+  // stubs (such that a common module isn't bundled twice).
+  await build({
```

Comment:
I believe in order for things to work with Deno, we need to prefix all Node.js stdlib imports with `node:`, so Deno knows that it should use the Node compat library. Does ESBuild do that for us? I don't think we have a unit test for it, but I think we should.

Reply:
Great call out. Sadly, esbuild can't do this for us automatically because it doesn't rewrite imports, but we can do that ourselves using an import map. See https://github.com/netlify/edge-bundler/pull/454/commits/99b95373e277c5f7428644fea4b07bec4c8b570b.